concent_api/core/message_handlers.py

```diff
@@ -1242,49 +1243,50 @@ def store_or_update_subtask(
     subtask_results_rejected: Optional[message.tasks.SubtaskResultsRejected] = None,
     force_get_task_result: Optional[message.concents.ForceGetTaskResult] = None,
 ) -> Subtask:
-    try:
-        subtask = Subtask.objects.get(
-            subtask_id = subtask_id,
-        )
-    except Subtask.DoesNotExist:
-        subtask = None
-
-    if subtask is not None:
-        if task_to_compute is not None and subtask.task_to_compute is not None:
-            validate_all_messages_identical([
-                task_to_compute,
-                deserialize_message(subtask.task_to_compute.data.tobytes()),
-            ])
-        subtask = update_subtask(
-            subtask=subtask,
-            state=state,
-            next_deadline=next_deadline,
-            set_next_deadline=set_next_deadline,
-            task_to_compute=task_to_compute,
-            report_computed_task=report_computed_task,
-            ack_report_computed_task=ack_report_computed_task,
-            reject_report_computed_task=reject_report_computed_task,
-            subtask_results_accepted=subtask_results_accepted,
-            subtask_results_rejected= subtask_results_rejected,
-            force_get_task_result=force_get_task_result,
-        )
-    else:
-        subtask = store_subtask(
-            task_id=task_id,
-            subtask_id=subtask_id,
-            provider_public_key=provider_public_key,
-            requestor_public_key=requestor_public_key,
-            state=state,
-            next_deadline=next_deadline,
-            task_to_compute=task_to_compute,
-            report_computed_task=report_computed_task,
-            ack_report_computed_task=ack_report_computed_task,
-            reject_report_computed_task=reject_report_computed_task,
-            subtask_results_accepted=subtask_results_accepted,
-            subtask_results_rejected=subtask_results_rejected,
-            force_get_task_result=force_get_task_result,
-        )
-    return subtask
+    with transaction.atomic(using='control'):
```

Comment:
There's a big problem here. You have recently turned my attention to the fact that a savepoint gets rolled back when the outer transaction is rolled back. So using nested `@atomic` does no guarantee that the changes will remain in the database.

Unfortunately I don't see any workaround other than doing two separate transactions. You'll have to use [`@non_atomic_requests`](https://docs.djangoproject.com/en/1.11/topics/db/transactions/#django.db.transaction.non_atomic_requests) on the views and create two non-nested `@atomic` blocks. The first one should create the `Subtask`. The other should do all the other operations.